frontend/src/metabase/visualizations/components/settings/ChartSettingsTableFormatting.jsx

```diff
@@ -92,106 +91,131 @@ const INPUT_CLASSNAME = "mt1 full";
-                columns: cols.length === 1 ? [cols[0].name] : [],
-              },
-              ...value,
-            ]);
-            this.setState({ editingRule: 0, editingRuleIsNew: true });
-          }}
-          onRemove={index => {
-            onChange([...value.slice(0, index), ...value.slice(index + 1)]);
-            MetabaseAnalytics.trackStructEvent(
-              "Chart Settings",
-              "Table Formatting",
-              "Remove Rule",
-            );
-          }}
-          onMove={(from, to) => {
-            const newValue = [...value];
-            newValue.splice(to, 0, newValue.splice(from, 1)[0]);
-            onChange(newValue);
-            MetabaseAnalytics.trackStructEvent(
-              "Chart Settings",
-              "Table Formatting",
-              "Move Rule",
-            );
-          }}
-        />
-      );
-    }
-  }
-}
-
-const SortableRuleItem = SortableElement(({ rule, cols, onEdit, onRemove }) => (
-  <RulePreview rule={rule} cols={cols} onClick={onEdit} onRemove={onRemove} />
-));
+  const { value, onChange, cols, canHighlightRow } = props;
 
-const SortableRuleList = SortableContainer(
-  ({ rules, cols, onEdit, onRemove }) => {
+  if (editingRule !== null && value[editingRule]) {
     return (
-      <div>
-        {rules.map((rule, index) => (
-          <SortableRuleItem
-            key={`item-${index}`}
-            index={index}
-            rule={rule}
-            cols={cols}
-            onEdit={() => onEdit(index)}
-            onRemove={() => onRemove(index)}
-          />
-        ))}
-      </div>
+      <RuleEditor
+        canHighlightRow={canHighlightRow}
+        rule={value[editingRule]}
+        cols={cols}
+        isNew={editingRuleIsNew}
+        onChange={rule => {
+          onChange([
+            ...value.slice(0, editingRule),
+            rule,
+            ...value.slice(editingRule + 1),
+          ]);
+        }}
+        onRemove={() => {
+          onChange([
+            ...value.slice(0, editingRule),
+            ...value.slice(editingRule + 1),
+          ]);
+          setEditingRule(null);
+          setEditingRuleIsNew(null);
+        }}
+        onDone={() => {
+          setEditingRule(null);
+          setEditingRuleIsNew(null);
+        }}
+      />
     );
-  },
-);
+  } else {
+    return (
+      <RuleListing
+        rules={value}
+        cols={cols}
+        onEdit={index => {
+          setEditingRule(index);
+          setEditingRuleIsNew(false);
+        }}
+        //This needs to be
+        onAdd={async () => {
+          await onChange([
+            {
+              ...DEFAULTS_BY_TYPE["single"],
+              // if there's a single column use that by default
+              columns: cols.length === 1 ? [cols[0].name] : [],
+              id: value.length,
+            },
+            ...value,
+          ]);
+          setEditingRuleIsNew(true);
+          setEditingRule(0);
+        }}
```

Comment:
Why does this function need to be async? It looks like you typed half of your thought in that comment :laughing:

Reply:
I'm easily distracted 🤦 

The logic here is kinda weird, but this needs to be async because otherwise value isn't updated by the time we do the next render. The end result is that we don't automatically pop open the column select for you if you click "Add a rule" and you already have an existing rule.